### PR TITLE
Enabled --minimal-modules.

### PR DESCRIPTION
**Pull request by vasslitvinov**
Pretty straightforward. Includes a bug fix in compiler.

For things I put in ChapelBase.chpl, my motivation was to get strings,
c_strings and ints working. I tossed in reals for a good measure, too.
Not sure how much of tuple stuff is needed. With my upcoming varargs PR,
hopefully these can be taken out altogether.

For minModIntRealString.chpl, I used "%ld" to printf-out Chapel's "int".
I hope this works on all platforms.

A note on string literals:

ChapelBase.chpl contains a string literal - returned by _defaultOf(string).
If we take it out and user code contains no string literals, then
the compiler will not be generating chpl__initStringLiterals() and
things will stop working. To fix them:

* Add the following e.g. to ChapelBase.chpl:
    extern proc chpl__initStringLiterals() {}
    chpl__initStringLiterals();

* In functionResolution.cpp, guard this:

    resolveFns(initStringLiterals);

with "if (initStringLiterals)".

### CODE REVIEW

**Comment by vasslitvinov** [comment]
@bradcray would you review?

**Comment by vasslitvinov** [comment]
P.S.1. Turns out I reformatted initCopy and autoDestroy in ChapelBase.chpl. This is not essential, just this way they use less space.

P.S.2. A few functions in ChapelBase.chpl do not have the restrictions that their counterparts in non-minimal modules have. This includes initCopy et al., cast, assignment. E.g assignment simply invokes primitive("=") regardless of argument types. This will work in most cases not explicitly anticipated, however will break sometimes, e.g. when assigning between string and c_string.

P.S.3. I switched __primitive calls to use c_strings instead of strings. This might lead to simpler IR, however should not be necessary.

File: test/compflags/bradc/minimalModules/minModIntRealString.chpl [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+var i: int;
+var r: real;
+var str = "Hello";
+var cs  = c"world";
+
+i = 5;
+r = 6.6;
+i = i + 7;
+
+printf("%ld %g  %s, %s!\n", i, r, str: c_string, cs);
```
Comment (bradcray) [inline test/compflags/bradc/minimalModules/minModIntRealString.chpl]: The %ld fails on Mac using clang...  :(

     /var/folders/69/jbk0y4q95lq8xhhwpqgwf_rh000bxs/T//chpl-bradc-73169.deleteme/minModIntRealString.c:19:29: error: format specifies type 'long' but the argument has type 'int64_t' (aka 'long long') [-Werror,-Wformat]
    > printf("%ld %g  %s, %s!\n", i_chpl, r_chpl, ret_chpl, cs_chpl);
    >         ~~~                 ^~~~~~
    >         %lld
    > 1 error generated.

File: compiler/main/driver.cpp [hunk 2]
```diff
@@ -793,7 +793,7 @@ static ArgumentDescription arg_desc[] = {
  {"remove-unreachable-blocks", ' ', NULL, "[Don't] remove unreachable blocks after resolution", "N", &fRemoveUnreachableBlocks, "CHPL_REMOVE_UNREACHABLE_BLOCKS", NULL},
 
  // Disabled until fix is in place for strings being defined in the modules
```
Comment (bradcray) [inline compiler/main/driver.cpp]: This "disabled" comment needs to be removed.

**Comment by bradcray** [comment]
What philosophy did you take in determining what should or should not be put into the minimal ChapelBase.chpl?  E.g., my original philosophy was to put the minimal amount in that would permit a trivial Chapel program to compile and no more.  I get the sense that you've been a bit more generous in your approach, but don't know what guided the decision to include, or not include, specific things.

**Comment by vasslitvinov** [comment]
@bradcray Let's stick with your original philosophy and define a trivial Chapel program to be one that uses, for example:

* int - incl. declare variables of this type and assign among them
* string - ditto
* varargs (for printf)
* function definitions and calls with the above
* maybe also int(8) so we can write portable tests

I also suggest having tests in your test directory that exercise these - roughly what's there right now - then trim ChapelBase.chpl just so it lets these tests run.

**Comment by bradcray** [comment]
Hi Vass -- I definitely prefer that style of philosophy -- particularly if it clears up your PS#2 which seemed like it might be related (though I'm not confident) and seemed like it could be a stumbling block for developers who ran afoul of it.  Specifically, my approach was really along the lines of "You want to be able to do this thing that seems quite primitive but is actually defined by modules?  Then define what you need to in order to do so (e.g., by cutting and pasting in from normal modules)."  What I couldn't tell from your PR was how much of the new string story required these additions (given that strings are used fairly directly by the compiler I think -- or so Kyle reported...) which is why I asked.

w.r.t. your specific list of features:
* int I think is a good basic starting point and your point about int(8) sounds good as well -- I think I like your inclusion of _defaultOf() for these types in ChapelBase.chpl even though one _could_ write a program without them by including them in the user code...

* string I think could definitely be left out if we could get by with c_strings -- but again, perhaps the compiler's internal use of strings necessitates this now for some reason (?)

* varargs: if this added more features to the minimal modules, I could imagine requiring the developer to declare their 1-, 2-, 3-arg overload versions as they needed them rather than relying on varargs.  Although I see that I used/relied on varargs in my original tests...  What module features do these exercise?  What could we cut out if we didn't have them on this list?

I'll be really happy to see this come back online again!  I think it's a valuable developer feature even though I haven't had a chance to use it myself in awhile.

**Comment by vasslitvinov** [comment]
I can leave string unsupported. That will require minor mods to have the compiler create chpl__initStringLiterals() when there are no string literals in the program.

Looking forward, how do you feel providing additional modules under modules/minimal that users can "use" to make additional features available?

E.g. use String; ==> they can use strings with --minimal-modules. Ditto reals, some arith ops like equality comparison, perhaps even tasks and foralls? I can see myself add these as I create them for my development tasks. The String module would simply contain the stuff that's currently in ChapelBase.

**Comment by bradcray** [comment]
I like that approach -- basically the minimal amount is 'use'd by default, but there are helper modules you can 'use' to get certain features.  I agree that these should be introduced lazily as needed rather than worrying about it too much now.

In the case of strings, I could definitely see supporting them by default due to the need for minor compiler mods (i.e., getting minimal modules working is more important to me than precisely what is in the minimal modules)  -- or worse, if those mods will make it hard for developers to incorporate strings into their code manually or via 'use String' as you suggest with --minimal-modules set.

I remain unclear on what can be done, if anything, to relieve the PS#2 issue.  Would removing the default assignment in ChapelBase reduce surprises/gotchas (while disabling assignment of course...)?

**Comment by vasslitvinov** [comment]
Brad -

* I will make those minor compiler mods - they are worth having IMO.

* This will allow me to not support strings by default, which I will also do.

* I will clear up my PS#2. It is straightforward - just restore the restrictions that I removed. It is not related to anything.

* I will provide int and int(8) support by default, with defaultOf and assignment.

* I will take vararg support out by default. I will probably provide a separate Tuples module instead to use if varargs are desired. Your examples included tuple stuff explicitly for varargs. How about converting your examples to fixed-arity printf overloads, in the spirit of minimality?

* How do you feel about creating test/modules/minimal/ for more min-mod tests, like tests that "use" String and other include-explicitly modules?

**Comment by bradcray** [comment]
> Your examples included tuple stuff explicitly for varargs. How about converting your examples to
> fixed-arity printf overloads, in the spirit of minimality?

Yeah, I agree with this.

> How do you feel about creating test/modules/minimal/ for more min-mod tests, like tests that "use" 
> String and other include-explicitly modules?

Sounds good to me.

The rest sounds good as well, thanks.

**Comment by vasslitvinov** [comment]
@bradcray how do you like it now?

**Comment by bradcray** [comment]
This looks good, thanks.  You'll need to update your commit message w.r.t. the way the patch ended up, and if you could explain the motivation behind some of the refactorings you did (which I understand, but a future reader might not) in the revised message, that'd be ideal.

**Pull request merged**